dali/pipeline/operators/detection/random_crop.cc

```diff
@@ -0,0 +1,347 @@
+    out_idx += (bounds[2] - bounds[0]) * C;
+  }
+}
+
+}  // namespace detail
+
+template <>
+void SSDRandomCrop<CPUBackend>::RunImpl(SampleWorkspace *ws, const int idx) {
+  // [H, W, C], dtype=uint8_t
+  const auto& img = ws->Input<CPUBackend>(0);
+  // [N] : [ltrb, ... ], dtype=float
+  const auto& bboxes = ws->Input<CPUBackend>(1);
+  const auto& labels = ws->Input<CPUBackend>(2);
+
+  auto N = bboxes.dim(0);
+  const float* bbox_data = bboxes.data<float>();
+
+  const int* label_data = labels.data<int>();
+
+  // [1x4]
+  Tensor<CPUBackend> crop_attempt;
+  crop_attempt.Resize({1, 4});
+  float *crop_ptr = crop_attempt.mutable_data<float>();
+
+  // printf("bbox_data: \n");
+  // for (int i = 0; i < N; ++i) {
+    // const float* b = &bbox_data[i * 4];
+    // printf("[%f, %f, %f, %f],\n", b[0], b[1], b[2], b[3]);
+  // }
+
+  // iterate until a suitable crop has been found
+  while (true) {
+    auto opt_idx = int_dis_(gen_);
+    auto option = sample_options_[opt_idx];
+
+    // printf("sample %d : option number %d\n", ws->data_idx(), opt_idx);
+    if (option.no_crop()) {
+      // printf("sample %d : no crop\n", ws->data_idx());
+      // copy directly to output without modification
+      ws->Output<CPUBackend>(0)->Copy(img, 0);
+      ws->Output<CPUBackend>(1)->Copy(bboxes, 0);
+      ws->Output<CPUBackend>(2)->Copy(labels, 0);
+      // printf("Boxes - %ld\n", N);
+      return;
+    }
+
+    // input is HWC ordering
+    auto htot = img.dim(0);
+    auto wtot = img.dim(1);
+
+    auto min_iou = option.min_iou();
+
+    // make num_attempts_ tries to get a valid crop
+    for (int i = 0; i < num_attempts_; ++i) {
+      auto w = float_dis_(gen_);
+      auto h = float_dis_(gen_);
+
+      // aspect ratio check
+      if ((w / h < 0.5) || (w / h > 2.)) {
+        // printf("FAILED ASPECT CHECK (%f, %f)\n", w, h);
+        continue;
+      }
+
+      // need RNG generators for left, top
+      std::uniform_real_distribution<float> l_dis(0., 1. - w), t_dis(0., 1. - h);
+      auto left = l_dis(gen_);
+      auto top = t_dis(gen_);
+
+      auto right = left + w;
+      auto bottom = top + h;
+
+      crop_ptr[0] = left;
+      crop_ptr[1] = top;
+      crop_ptr[2] = right;
+      crop_ptr[3] = bottom;
+
+      // printf("crop: [%f, %f, %f, %f]\n", left, top, right, bottom);
+
+      // returns ious : [N, M]
+      Tensor<CPUBackend> ious = detail::cpu_iou(bboxes, crop_attempt);
+      const float *ious_data = ious.data<float>();
+
+      // make sure all the calculated IoUs are in the range (min_iou, max_iou)
+      // Note: ious has size N*M, but M = 1 in this case
+      bool fail = false;
+      for (int j = 0; j < N * 1; ++j) {
+        if (ious_data[j] <= min_iou) fail = true;
+      }
+      // generate a new crop
+      if (fail) {
+        // printf("FAILED IoU CHECK\n");
+        continue;
+      }
+
+      // discard any bboxes whose center is not in the cropped image
```

Comment:
Is there any reason why this is done after calculating IoU? As far as I see, you could potentially discard all bounding boxes and save yourself the trouble of computing IoU if these two steps where swapped in their ordering

Reply:
Because if any `IoU(bbox, proposed_crop <= min_IoU` we discard and generate a new proposed crop. What you're suggesting is a change in the algorithm. This may still be valid, but it's not the same

Reply:
I see the logic in the code, however the volume of work in IoU heavily outweights the selection by centroid. I'd jump at the chance to save ourselves few (or all) redundant IoU calculations... I will benchmark that in our refactor

Reply:
I think what @pribalta is suggesting is to fuse IoU with checking ```proposed_crop <= min_IoU ```, then if you encounter the first bbox that fails the check you can skip calculating next one and generate a new set. Now we are generating full set even if the first value is not meeting our threshold.
I don't think it changes logic in any way, as instead of couple for loops inside the cpu_iou, you can have big one with early exit on ```proposed_crop <= min_IoU ``` condition.
Anyway let us leave it as it is, we will return to this in 0.5.